### DISCUSSION TITLE
MPI vs serial displacement results

### DISCUSSION THREAD

**Discussion by Flolaffel** [idaholab/moose#28120] [Q&A Modules: Solid mechanics]
Hello,

I recently added MPI communication to my app and ran a few tests to ensure that my MPI runs deliver the same results as my serial runs. But they didn't. So I started debugging and what I found was that my MPI communication works as intended but MOOSE delivers slightly differing nodal displacements for serial and MPI.

Take the follwing input as an example and run it in serial and then with MPI. The nodal displacements will be very close but not equal. This seems like a minor difference but for my application this error adds up over time which leads to significant differences in my end result.

```
nx = 2
ny = 2
E0 = 1
nu = 0.3

[GlobalParams]
  displacements = 'disp_x disp_y'
[]

[Mesh]
  [domain]
    type = GeneratedMeshGenerator
    dim = 2
    xmin = -5
    xmax = 5
    ymin = -5
    ymax = 5
    nx = ${nx}
    ny = ${ny}
  []
  [top_right]
    type = ExtraNodesetGenerator
    input = domain
    new_boundary = tr
    coord = '5 5 0'
  []
  [bottom_right]
    type = ExtraNodesetGenerator
    input = top_right
    new_boundary = br
    coord = '5 -5 0'
  []
  [top_left]
    type = ExtraNodesetGenerator
    input = bottom_right
    new_boundary = tl
    coord = '-5 5 0'
  []
  [bottom_left]
    type = ExtraNodesetGenerator
    input = top_left
    new_boundary = bl
    coord = '-5 -5 0'
  []
[]

[Physics]
  [SolidMechanics]
    [QuasiStatic]
      [all]
        strain = SMALL
        add_variables = true
        incremental = false
        generate_output = 'stress_xx stress_yy stress_xy strain_xx strain_yy strain_xy vonmises_stress'
      []
    []
  []
[]

[BCs]
  [no_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'bl br tl'
    value = 0
  []
  [no_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'bl'
    value = 0
  []
  [disp]
    type = DirichletBC
    variable = disp_x
    boundary = 'tl'
    value = 0.5
  []
[]

[Materials]
  [elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = ${E0}
    poissons_ratio = ${nu}
  []
  [compute_stress]
    type = ComputeLinearElasticStress
  []
[]

[Preconditioning]
  [SMP]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Steady
  solve_type = Newton
[]

[VectorPostprocessors]
  [disp_x]
    type = NodalValueSampler
    sort_by = id
    variable = disp_x
    contains_complete_history = true
    execution_order_group = 10
  []
[]

[Outputs]
  csv = true
[]
```

So my question is: Is this a known effect and if yes what is the reason for the differing displacement results in serial vs MPI?

**Comment by GiudGiud** [comment]
Hello

The solutions are not exactly the same in serial and parallel because the linear solvers use different preconditioners by default in serial and parallel. The effect is within the solver tolerances (both solutions are converged to tolerances, and the solution should be unique if the problem is well-posed).
If you want the solution to be exactly the same you can look in the PETSc documentation for serial preconditioners like some LU.
I don't know if there are scalable solvers that also give the exact same results in serial

Guillaume

**Comment by Flolaffel** [comment]
Thank you for your answer.

> and the solution should be unique if the problem is well-posed

What do you mean by unique here? What are the criteria to call a solution "unique"? Indentical to the tolerance or completely indentical even at the n-th decimal outside the tolerance?

> If you want the solution to be exactly the same you can look in the PETSc documentation for serial preconditioners like some LU.

I'm not sure if understand that correctly. Should the serial and parallel solution be _exactly_ the same when using LU? Because it isn't.

**Reply by GiudGiud** [reply]
> What do you mean by unique here? 

Unique in a mathematical way. The finite element approximation of the true solution should be unique with the discretization parameters chosen. 

>  Indentical to the tolerance or completely indentical even at the n-th decimal outside the tolerance?

Close to identical to the tolerance. But there are many caveats. The relative and absolute tolerances behave differently, and the tolerance is in terms of residuals not in terms of how converged each DOFs is.

> Should the serial and parallel solution be exactly the same when using LU? Because it isn't.

Should be close given that the preconditioning is supposed to be serial. Maybe not all implementations are serial. You need to look at the PETSc documentation, I don't have all the specs of all PCs in mind
